### PR TITLE
fix data race in kubelet volume test: add lock for ut

### LINKED ISSUE #104057: The test `TestVolumeUnmountAndDetachControllerDisabled` test if failing from `kubelet_volumes_test.go`
#### Which jobs are failing:
pull-kubernetes-unit

#### Which test(s) are failing:
TestVolumeUnmountAndDetachControllerDisabled

#### Since when has it been failing:
It failed on `07/24/2021` but after that the test is passing

#### Testgrid link:
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/103631/pull-kubernetes-unit/1418992044973494272#1:build-log.txt%3A2386

#### Reason for failure:
DATA RACE

It seems that the data race is caused by below code

https://github.com/kubernetes/kubernetes/blob/5be21c50c269fc1d28e0bd31ab9dcb572ae7fac5/pkg/kubelet/kubelet_volumes_test.go#L317-L320

#### Anything else we need to know:
It failing of test is Flaky, it was not reproducible on my local machine too, I have used below command to run the test
```
sudo make test WHAT="./pkg/kubelet"  GOFLAGS="-v -count=1" KUBE_TEST_ARGS='-run ^TestVolumeUnmountAndDetachControllerDisabled$'
```

output on my local machine 

```
+++ [0801 06:08:49] Running tests without code coverage and with -race
=== RUN   TestVolumeUnmountAndDetachControllerDisabled
W0801 06:08:55.941792    4176 mutation_detector.go:53] Mutation detector is enabled, this will result in memory leakage.
I0801 06:08:55.942831    4176 desired_state_of_world_populator.go:146] "Desired state populator starts to run"
I0801 06:08:55.943127    4176 volume_manager.go:291] "Starting Kubelet Volume Manager"
E0801 06:08:55.945783    4176 reflector.go:138] k8s.io/client-go/informers/factory.go:134: Failed to watch *v1.CSIDriver: unhandled watch: testing.WatchActionImpl{ActionImpl:testing.ActionImpl{Namespace:"", Verb:"watch", Resource:schema.GroupVersionResource{Group:"storage.k8s.io", Version:"v1", Resource:"csidrivers"}, Subresource:""}, WatchRestrictions:testing.WatchRestrictions{Labels:labels.internalSelector(nil), Fields:fields.andTerm{}, ResourceVersion:""}}
I0801 06:08:56.145474    4176 reconciler.go:244] "operationExecutor.AttachVolume started for volume \"vol1\" (UniqueName: \"fake/fake-device\") pod \"foo\" (UID: \"12345678\") "
I0801 06:08:56.145588    4176 reconciler.go:157] "Reconciler: start to sync state"
I0801 06:08:56.145710    4176 operation_generator.go:369] AttachVolume.Attach succeeded for volume "vol1" (UniqueName: "fake/fake-device") from node "127.0.0.1" 
I0801 06:08:56.246900    4176 operation_generator.go:587] MountVolume.WaitForAttach entering for volume "vol1" (UniqueName: "fake/fake-device") pod "foo" (UID: "12345678") DevicePath "/dev/vdb-test"
I0801 06:08:56.247005    4176 operation_generator.go:597] MountVolume.WaitForAttach succeeded for volume "vol1" (UniqueName: "fake/fake-device") pod "foo" (UID: "12345678") DevicePath "/dev/sdb"
I0801 06:08:56.247084    4176 operation_generator.go:630] MountVolume.MountDevice succeeded for volume "vol1" (UniqueName: "fake/fake-device") pod "foo" (UID: "12345678") device mount path ""
I0801 06:08:56.548217    4176 reconciler.go:196] "operationExecutor.UnmountVolume started for volume \"vol1\" (UniqueName: \"fake/fake-device\") pod \"12345678\" (UID: \"12345678\") "
I0801 06:08:56.548267    4176 operation_generator.go:866] UnmountVolume.TearDown succeeded for volume "fake/fake-device" (OuterVolumeSpecName: "vol1") pod "12345678" (UID: "12345678"). InnerVolumeSpecName "vol1". PluginName "fake", VolumeGidValue ""
I0801 06:08:56.649605    4176 reconciler.go:312] "operationExecutor.UnmountDevice started for volume \"vol1\" (UniqueName: \"fake/fake-device\") on node \"127.0.0.1\" "
I0801 06:08:56.649609    4176 operation_generator.go:974] UnmountDevice succeeded for volume "vol1" %!(EXTRA string=UnmountDevice succeeded for volume "vol1" (UniqueName: "fake/fake-device") on node "127.0.0.1" )
I0801 06:08:56.750200    4176 reconciler.go:333] "operationExecutor.DetachVolume started for volume \"vol1\" (UniqueName: \"fake/fake-device\") on node \"127.0.0.1\" "
I0801 06:08:56.750269    4176 operation_generator.go:484] DetachVolume.Detach succeeded for volume "vol1" (UniqueName: "fake/fake-device") on node "127.0.0.1" 
I0801 06:08:56.844405    4176 volume_manager.go:297] "Shutting down Kubelet Volume Manager"
--- PASS: TestVolumeUnmountAndDetachControllerDisabled (0.91s)
PASS
ok      k8s.io/kubernetes/pkg/kubelet   1.052s
```

I would like to work on this issue, is it possible to reproduce this issue locally?

### PR DESCRIPTION
**Pull request by pacoxu**
/kind flake
Fixes #104057 

```release-note
NONE
```

### CODE REVIEW

File: pkg/kubelet/pod_workers_test.go [hunk 1]
```diff
@@ -105,6 +105,11 @@ func (f *fakePodWorkers) ShouldPodRuntimeBeRemoved(uid types.UID) bool {
 	defer f.statusLock.Unlock()
 	return f.removeRuntime[uid]
 }
+func (f *fakePodWorkers) SetPodRuntimeBeRemoved(uid types.UID) {
```
Comment (manugupt1) [inline pkg/kubelet/pod_workers_test.go:108]: The PodWorkers interface does not set `SetPodRuntimeBeRemoved`. 
Does it make sense for this to be a private implementation?

```
f.setPodRuntimeBeRemoved
```

Another question for my own understanding; is what happens in the original implementation? Does this wait for volumes to be unmounted (I understand its in the comments, but wanted to double check).

Comment (pacoxu) [reply pkg/kubelet/pod_workers_test.go]: To fix it gracefully, I think we need to use something like `waitForVolumeUnmount` to wait for fake unmount.

`SetPodRuntimeBeRemoved` is more like `unmount runtime by test itself`

Comment (pacoxu) [reply pkg/kubelet/pod_workers_test.go]: Updated.

**Review by adisky — COMMENTED** [review]
this should fix the race condition
/lgtm
```
WARNING: DATA RACE
Write at 0x00c00080eae8 by goroutine 113:
  k8s.io/kubernetes/pkg/kubelet.TestVolumeUnmountAndDetachControllerDisabled()
      /home/sharmaad/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet_volumes_test.go:319 +0xc73
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1193 +0x202

Previous read at 0x00c00080eae8 by goroutine 132:
  k8s.io/kubernetes/pkg/kubelet.(*fakePodWorkers).ShouldPodRuntimeBeRemoved()
      /home/sharmaad/go/src/k8s.io/kubernetes/pkg/kubelet/pod_workers_test.go:106 +0xbc
```

**Comment by pacoxu** [comment]
/assign @mrunalp 
for approval

**Comment by aojea** [comment]
/lgtm
Thanks

**Review by SergeyKanzhelev — APPROVED** [review]
/lgtm

**Comment by manugupt1** [comment]
/assign mrunalp 
as it has enough lgtm

**Pull request merged**